MVP implementation of error mode STOP/RESUME

simonbasle opened the pull request:
See #629, cc @dfeist. Only taken into account by `map` and `filter` for now.

simonbasle commented:
@smaldini this is now rebased on top of #830, to be rebased against master once #830 is merged.

dfeist commented on reactor-core/src/main/java/reactor/core/publisher/OnNextFailureStrategy.java:
```diff
@@ -0,0 +1,193 @@
+	 * hook and the incriminating source value is passed to the {@link Operators#onNextDropped(Object, Context)}
+	 * hook, allowing the sequence to continue with further values.
+	 */
+	OnNextFailureStrategy RESUME_DROP = new OnNextFailureStrategy() {
```
I'd just use `RESUME` personally, because if the event gets dropped or not really depends on what the user does with it in the callbacks. If they send it to a DLQ for example, then it hasn't been dropped.  These strategy names should reflect the behaviour from the Flux perspective IMO.

dfeist replied:
Ah, i see what you've done now, this is a built-in version of RESUME that just drop.  Needs a bit of cleanup because the strategy isn't really a strategy, and exposes constants/methods to implement different strategies.

dfeist commented on reactor-core/src/main/java/reactor/core/publisher/OnNextFailureStrategy.java:
```diff
@@ -0,0 +1,193 @@
+ *
+ * @author Simon Baslé
+ */
+public interface OnNextFailureStrategy {
```
Is the use of `failure` in this PR instead of `error` deliberate, just wondering because `error` is used everywhere else.

smaldini commented on reactor-core/src/main/java/reactor/core/publisher/FluxFilter.java:
```diff
@@ -206,7 +215,13 @@ public void onNext(T t) {
 			}
 			catch (Throwable e) {
-				onError(Operators.onOperatorError(s, e, t, actual.currentContext()));
+				Throwable e_ = Operators.onNextFailure(t, e, actual.currentContext(), s);
```
onNextFailure should encapsulate the whole thing ( if then cancel+error else request ) being passed a CoreSubscriber.

smaldini commented on reactor-core/src/main/java/reactor/core/publisher/FluxFilter.java:
```diff
@@ -230,7 +245,10 @@ public boolean tryOnNext(T t) {
 			}
 			catch (Throwable e) {
-				onError(Operators.onOperatorError(s, e, t, actual.currentContext()));
+				Throwable e_ = Operators.onNextFailure(t, e, actual.currentContext(), s);
```
same thing that previous comment adapted for ConditionalSubscriber

simonbasle replied:
I did it that way because some operators slightly deviate from a "standard" actual.onError implementation. For instance, consider `FluxMap`:

```java
Throwable e_ = Operators.onNextError(t, e, actual.currentContext(), s);
				if (e_ != null) {
					done = true;
					actual.onError(e_);
					return true;
				}
				else {
					return false;
				}
```

smaldini commented on reactor-core/src/main/java/reactor/core/publisher/FluxFilterFuseable.java:
```diff
@@ -183,21 +191,43 @@ public T poll() {
+							throw e_;
+						}
+						else {
+							return poll();
```
dangerous recursive poll, should continue;

smaldini commented on reactor-core/src/main/java/reactor/core/publisher/OnNextFailureStrategy.java:
```diff
@@ -0,0 +1,246 @@
+ *
+ * @author Simon Baslé
+ */
+public interface OnNextFailureStrategy {
```
No strategy for now, BiFunction is fine

simonbasle commented:
(rebased on current master)

simonbasle commented:
rebased on master, addressed comments by @smaldini on recursive `poll()` (e045107) and hid `OnNextFailureStrategy` from public API (0f006dc).

At `Hooks` level the strategy is set up as a `BiFunction`. It can be an intersection of `BiFunction` and `BiPredicate`. The later allows to conditionally apply the strategy, with a fallback to `onOperatorError`. The strategy interface is kept internally as it allows to add a 3rd parameter (the `Context`) to processing.

Next step would be to expose factory methods of `OnNextFailureStrategy` publicly, albeit returning a bland `BiFunction`. Internally these would still be `OnNextFailureStrategy` so they can have a predicate and work with the `Context`.

dfeist commented on reactor-core/src/main/java/reactor/core/publisher/OnNextFailureStrategy.java:
```diff
@@ -0,0 +1,274 @@
+	 * It must deal with potential {@code null}s.
+	 * @return a new {@link OnNextFailureStrategy} that allows resuming the sequence.
+	 */
+	static OnNextFailureStrategy resume(Consumer<Throwable> errorConsumer,
```
@simonbasle Why not a `BiConsumer` here? Especially given the two are so related given the error is caused by onNext of the value.

simonbasle replied:
That makes sense. I think at some point the dropping implementation was reusing the consumer implementation, so having separate consumers would allow to detect the `value == null` case and skip the value consumer entirely, making the reuse as simple as calling `Operators::onNextDropped`.

dfeist replied:
Maybe, but then it's a significant issue if/when you need to use both even if it's a simple as log a message referencing error and value.  Any issue with switching this to BiConsumer, and doing the adaption to next dropped consumer internally?

dfeist commented on reactor-core/src/main/java/reactor/core/publisher/OnNextFailureStrategy.java:
```diff
@@ -0,0 +1,274 @@
+	 */
+	static OnNextFailureStrategy resumeIf(
+			Predicate<Throwable> causePredicate,
+			Consumer<Throwable> errorConsumer,
```
Idem

dfeist replied:
@simonbasle Also wondering if there is any use for context in error callbacks.

dfeist commented on reactor-core/src/main/java/reactor/core/publisher/OnNextFailureStrategy.java:
```diff
@@ -0,0 +1,274 @@
+		@Nullable
+		public Throwable process(Throwable error, @Nullable Object value, Context context) {
+			if (errorPredicate == null) {
+				Exceptions.throwIfFatal(error);
```
Nice that it's possible to match specific fatal errors with predicate and skip the default `throwIfFatal` :-)

dfeist commented on reactor-core/src/main/java/reactor/core/publisher/Operators.java:
```diff
@@ -376,6 +377,109 @@ public static RuntimeException onRejectedExecution(Throwable original, Context c
+		return strategy;
+	}
+
+	public static final BiFunction<? super Throwable, Object, ? extends Throwable> onNextErrorFunction(Context context) {
```
This still need tying in?

simonbasle replied:
yes, we don't want to expose a new interface for now, but the interface allows for smarter checks for internal implementations when it is used as a `BiPredicate`.

dfeist replied:
Also, looks like this would be the way to support context in the future too..

dfeist commented on reactor-core/src/main/java/reactor/core/publisher/OnNextFailureStrategy.java:
```diff
@@ -0,0 +1,274 @@
+			this.delegateProcessor = delegateProcessor;
+			if (delegateProcessor instanceof BiPredicate) {
+				//noinspection unchecked
+				this.delegatePredicate = (BiPredicate<? super Throwable, Object>) delegateProcessor;
```
Is it not best to pass these separately? Otherwise using this (at least with lambda's) is a pain.  

Or is there idea, that to pass lambda' you'd use the other version?  The main issue with the other handler is the two consumers rather than a biconsumer.

simonbasle replied:
The idea is to use the other versions, yeah. This one is just an adapter, that has additional hidden logic for `BiFunction`s that are also `BiPredicate` (a sneaky way of allowing user-provided equivalents of the `OnNextErrorStrategy` interface without exposing the interface).

dfeist replied:
Ok, so need `BiConsumer` in other version and this is for more advanced custom strategies.

simonbasle commented:
@dfeist just rebased on top of master (including doOnEach change with Context)

simonbasle commented:
fixed by ce62135, do include in 3.2.0 as part of #1023
